eta/core/video.py

```diff
@@ -108,6 +116,686 @@ def glob_videos(dir_):
+        return _attrs
+
+    def _ensure_frame(self, frame_number):
+        if not self.has_frame(frame_number):
+            self.frames[str(frame_number)] = VideoFrameLabels(frame_number)
+
+    def _validate_frame_labels(self, frame_labels):
+        if self.has_schema:
+            for frame_attr in frame_labels.attrs:
+                self.schema.validate_frame_attribute(frame_attr)
+            for obj in frame_labels.objects:
+                self.schema.validate_object(obj)
+
+    def _validate_frame_attribute(self, frame_attr):
+        if self.has_schema:
+            self.schema.validate_frame_attribute(frame_attr)
+
+    def _validate_object(self, obj):
+        if self.has_schema:
+            self.schema.validate_object(obj)
+
+    def _validate_schema(self):
+        if self.has_schema:
+            for frame_labels in itervalues(self.frames):
+                self._validate_frame_labels(frame_labels)
+
+    @classmethod
+    def from_dict(cls, d):
+        '''Constructs a VideoLabels from a JSON dictionary.'''
+        frames = OrderedDict(
+            (fn, VideoFrameLabels.from_dict(vfl))
+            for fn, vfl in iteritems(d["frames"])
+        )
+
+        schema = d.get("schema", None)
+        if schema is not None:
+            schema = VideoLabelsSchema.from_dict(schema)
+
+        return cls(frames=frames, schema=schema)
+
+
+class VideoLabelsSchema(Serializable):
+    '''A schema for a VideoLabels instance.'''
+
+    def __init__(self, frames=None, objects=None):
+        '''Creates a VideoLabelsSchema instance.
+
+        Args:
+            frames: an AttributeContainerSchema describing the frame attributes
+                of the video
+            objects: a dictionary mapping object labels to
+                AttributeContainerSchemas describing the object attributes of
+                each object class
+        '''
+        self.frames = frames or AttributeContainerSchema()
+        self.objects = defaultdict(lambda: AttributeContainerSchema())
+        if objects is not None:
+            self.objects.update(objects)
+
+    def add_frame_attribute(self, frame_attr):
+        '''Incorporates the given frame attribute into the schema.
+
+        Args:
+            frame_attr: an Attribute
+        '''
+        self.frames.add_attribute(frame_attr)
+
+    def add_frame_attributes(self, frame_attrs):
+        '''Incorporates the given frame attributes into the schema.
+
+        Args:
+            frame_attrs: an AttributeContainer of frame attributes
+        '''
+        self.frames.add_attributes(frame_attrs)
+
+    def add_object_label(self, label):
+        '''Incorporates the given object label into the schema.'''
+        self.objects[label]  # adds key to defaultdict
+
+    def add_object_attribute(self, label, obj_attr):
+        '''Incorporates the Attribute for the object with the given label
+        into the schema.
+        '''
+        self.objects[label].add_attribute(obj_attr)
```

Comment:
should be in a try:catch, no?  If label is not in objects, then boom.

Reply:
`objects` is a `defaultdict`, so it will automatically initialize as necessary. I use `defaultdict` heavily.